src/object.jl

```diff
@@ -372,3 +372,147 @@ function Base.read(io::PrefetchedDownloadStream, ::Type{UInt8})
+* `store::AbstractStore`: The S3 Bucket / Azure Container object
+* `key::String`: S3 key / Azure blob resource name
+
+# Keywords
+* `credentials::Union{CloudCredentials, Nothing}=nothing`: Credentials object used in HTTP
+    requests
+* `concurrent_writes_to_channel::Int=(4 * Threads.nthreads())`: The number of concurrent
+    writes to the channel. Defaults to 4 times the number of threads. We use this value to
+    initialize a semaphore to perform throttling in case the writing to the channel is much
+    faster to uploading to blob storage.
+* `kwargs...`: HTTP keyword arguments are forwarded to underlying HTTP requests,
+
+## Examples
+```
+# Get an IO stream for a remote CSV file `test.csv` living in your S3 bucket
+io = MultipartUploadStream(bucket, "test.csv"; credentials)
+```
+
+## Performance
+We have benchmarked the performance of MultipartUploadStream for smaller (~39MB) and larger
+files up to ~860MB. For smaller files the performance is similar to an S3.put call, whereas
+for larger ones we do see a degradation of about 18% after ~300MB, that is growing more as
+the size of the uploaded file grows. We need to investirage further the cause of this.
+Some benchmark results can be found in https://github.com/JuliaServices/CloudStore.jl/pull/46#issuecomment-1804298709
+and https://github.com/JuliaServices/CloudStore.jl/pull/46#issuecomment-1810558208.
+"""
+mutable struct MultipartUploadStream{T <: AbstractStore} <: IO
+    store::T
+    url::String
+    credentials::Union{Nothing, AWS.Credentials, Azure.Credentials}
+    uploadState::Union{String, Nothing}
+    eTags::Vector{String}
+    upload_queue::Channel{Tuple{Int, Vector{UInt8}}}
+    cond_wait::Threads.Condition
+    cur_part_id::Int
+    ntasks::Int
+    exc::Union{Exception, Nothing}
+    sem::Base.Semaphore
+
+    function MultipartUploadStream(
+        store::T,
+        key::String;
+        credentials::Union{CloudCredentials, Nothing}=nothing,
+        concurrent_writes_to_channel::Int=(4 * Threads.nthreads()),
+        kw...
+    ) where {T<:AbstractStore}
+        url = makeURL(store, key)
+        uploadState = API.startMultipartUpload(store, key; credentials, kw...)
+        io = new{T}(
+            store,
+            url,
+            credentials,
+            uploadState,
+            Vector{String}(),
+            Channel{Tuple{Int, Vector{UInt8}}}(Inf),
+            Threads.Condition(),
+            0,
+            0,
+            nothing,
+            Base.Semaphore(concurrent_writes_to_channel)
+        )
+        return io
+    end
+end
+
+function Base.write(io::MultipartUploadStream, bytes::Vector{UInt8}; kw...)
+    local part_n
+    Base.@lock io.cond_wait begin
+        io.ntasks += 1
+        io.cur_part_id += 1
+        part_n = io.cur_part_id
+        notify(io.cond_wait)
+    end
+    Base.acquire(io.sem)
+    # We expect the data chunks to be written in order in the channel.
+    put!(io.upload_queue, (part_n, bytes))
+    Base.release(io.sem)
```

Comment:
The release should happen after the task is done uploading the chunk, otherwise this wouldn't block